### PR TITLE
Benchmarks: discover v2 Node Renderer bundles

### PR DESCRIPTION
**Pull request by justin808**
## Why

The unattended 05:00 HST run associated with closed issue #4073 reached Pro Node Renderer setup at merge commit `65a8269ed41ddba0b0ec351f1496aef6e5fb9124`, but its benchmark harness reported that no production bundles existed.

The cache was not empty. Pro pre-seeding had staged canonical v2 artifact directories named `rorp-v2-s-<64 hex>` and `rorp-v2-r-<64 hex>`, while `bench-node-renderer.rb` still accepted only legacy `<hex>-production` names. That matcher returned zero entries and produced the misleading empty-bundle failure.

This is follow-up work for #4073 and PR #4853. It intentionally does not reopen or auto-close the already-completed issue.

## What changed

- Recognize only the strict canonical v2 server and RSC artifact names.
- Select the newest candidate independently for each encoded role, preventing persistent-cache leftovers from being benchmarked as current output.
- Fail closed when same-role candidates share the exact newest mtime instead of choosing one nondeterministically.
- Validate each selected candidate's exact `<id>/<id>.js` payload and fail closed if the newest artifact is incomplete; never fall back to an older readable artifact.
- Preserve newest-first legacy discovery only when no v2 candidates exist.
- Add regression coverage for v2 roles, malformed names, persistent stale entries, fail-closed payload validation, and legacy fallback.

No Pro runtime, scheduler, workflow, dependency, lockfile, or credential handling changed. No Pro license was requested or exposed.

## Validation

- `bundle exec rspec benchmarks/spec/bench_node_renderer_spec.rb` — 18 examples, 0 failures
- `bundle exec rspec benchmarks/spec` — 350 examples, 0 failures
- `.agents/bin/validate --changed` — passed docs sidebar, Ruby/benchmark RuboCop, package build, ESLint, Prettier, and type checking
- `.agents/bin/lint` — passed, including 260 runs / 2,258 assertions with no failures or errors
- `git diff --check` — clean
- Focused RuboCop — 2 files inspected, no offenses

## Review and churn

- The first independent Codex review found one real priority issue: accepting every v2 directory could benchmark stale persistent-cache artifacts. The implementation was narrowed to newest-per-role selection with fail-closed payload validation, and regression tests were added.
- The refined uncommitted review and the committed-diff review both returned clean.
- The first hosted review wave found one real P2 edge case: tied newest mtimes for one role made Ruby's `max_by` choice nondeterministic. The final review-fix commit now rejects that ambiguous state and adds a regression test. A low-risk hosted Claude DRY suggestion was batched into the same necessary commit by centralizing the strict role matcher.
- The final uncommitted review returned clean.
- The next hosted Claude wave found one real legacy-compatibility regression: eager payload validation changed the pre-existing legacy discovery contract. The review's supporting claim that the later live categorizer skips HTTP 410 responses was disproved during independent QA (`curl` does not use `--fail`), so this PR does not expand into that separate pre-existing behavior. The final review-fix confines exact-payload fail-closed validation to v2 discovery and adds a red/green example proving no eager validation was added to the legacy path.
- Independent read-only QA at exact head `922d693226c2ea7c305140b58b1eb9e66a609c98` returned CLEAN: focused finder specs 5/0, focused file 18/0, full benchmark specs 350/0, scoped RuboCop clean, and issue #4073 still CLOSED.
- The configured Claude Opus 4.8 review and simplify pass were unavailable because that provider reported its weekly limit, resetting Aug 10 at 21:00 HST. No result is inferred from the unavailable pass.
- Post-push fix cycles: 2; tied-mtime/DRY feedback was handled in `278f51f1a2ab8f45341ebe74b4d9604fac3394ad`, and legacy discovery compatibility was restored in `922d693226c2ea7c305140b58b1eb9e66a609c98`.

## Changelog

Not user-visible: this is a narrow internal benchmark-harness compatibility fix.

## Remaining operational receipt

The repository fix is not the final operational proof. After merge, the next unattended 05:00 HST run must record one exact-main receipt in which Core, Pro, and Pro Node Renderer all complete and upload. That receipt will be added durably while issue #4073 remains closed.

### QA Evidence

- QA lane: `rord-4073-qa-sol`, read-only QA in the implementation worktree, target claim covered by the batch claim; final heartbeat status `done`
- Scope checked: both changed benchmark files, exact-head discovery semantics, the persistent-cache failure shape, and closed issue state
- Tested at: `922d693226c2ea7c305140b58b1eb9e66a609c98` on base `f4a1bde1e880b5ea69c08c55f620fbf6c10afe19`
- Automated checks: focused finder specs 5/0; focused file 18/0; full benchmark specs 350/0; scoped RuboCop clean; `git diff --check` clean
- Manual checks: metadata-only audit found four canonical v2 directories and zero legacy-regex matches; a payload-less legacy fixture remained discoverable exactly as on the base branch; the separate legacy live-classification path was confirmed unchanged and left out of scope; no credential contents were read
- Findings: none after the stale-fallback finding was fixed and re-reviewed
- QA required: yes
- QA required rationale: unattended benchmark execution and persistent renderer-cache selection are operationally sensitive
- QA lane status: satisfied
- Release-blocking status: clear
- Process-gap disposition: script

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved bundle discovery for rendering benchmarks by selecting the newest valid bundles.
  * Added validation for missing or ambiguous bundle payloads, providing clearer failures.
  * Preserved compatibility with legacy production bundles when newer bundles are unavailable.

* **Tests**
  * Added coverage for bundle selection, ordering, invalid entries, missing payloads, and ambiguity handling.

## Merge readiness

Confidence note:
- Validated: focused benchmark file 18/0; full benchmark specs 350/0; `.agents/bin/validate --changed`; `.agents/bin/lint`; scoped RuboCop; `git diff --check`; exact-head required gate, hosted lint/build, benchmark specs, CodeQL, CodeRabbit, and Claude review all passed.
- Evidence: exact head `922d693226c2ea7c305140b58b1eb9e66a609c98`; full-history address-review summary https://github.com/shakacode/react_on_rails/pull/4858#issuecomment-5229087333; all four review threads resolved; strict `script/pr-merge-ledger` completed with no violations or unknown fields.
- UNKNOWN: none affecting merge safety.
- Residual risk: the merged `main` commit has not yet been exercised by the real unattended 05:00 HST benchmark; that known post-merge operational receipt remains required and is not waived.

Decision points: 0 during implementation and review processing.

Merge authority: explicit maintainer approval for exact head `922d693226c2ea7c305140b58b1eb9e66a609c98`; state: `authorized-for-guarded-merge`; durable receipt: https://github.com/shakacode/react_on_rails/pull/4858#issuecomment-5229271155.

### CODE REVIEW

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: benchmarks/bench-node-renderer.rb [hunk 1]
```diff
@@ -83,6 +83,33 @@ def add_to_summary(*parts)
...
 
+def bundle_entry_mtime(bundles_dir, entry)
+  File.mtime(File.join(bundles_dir, entry))
+end
+
+def newest_v2_bundles_by_role(bundles_dir, candidates)
+  candidates.group_by { |entry| entry[/\Arorp-v2-([sr])-/, 1] }
+            .values
+            .map do |role_entries|
+    role_entries.max_by { |entry| bundle_entry_mtime(bundles_dir, entry) }
```
[the remark this answers is not in the document]
Comment (justin808) [reply benchmarks/bench-node-renderer.rb]: Fixed in 278f51f1a2ab8f45341ebe74b4d9604fac3394ad. Selection now groups each role by exact directory mtime and fails closed when multiple candidates share the newest timestamp, with deterministically sorted names in the diagnostic. The new regression test reproduces the tied-mtime case; exact-head QA passed 17 focused-file examples and all 349 benchmark specs.

File: benchmarks/bench-node-renderer.rb [hunk 2]
```diff
@@ -83,6 +83,33 @@ def add_to_summary(*parts)
...
+  end
+end
+
+def validate_bundle_payloads!(bundles_dir, bundles)
+  bundles.each do |entry|
+    payload_path = File.join(bundles_dir, entry, "#{entry}.js")
+    next if File.file?(payload_path)
+
+    role = entry[/\Arorp-v2-([sr])-/, 1]
```
[the remark this answers is not in the document]
Comment (justin808) [reply benchmarks/bench-node-renderer.rb]: Handled in 278f51f1a2ab8f45341ebe74b4d9604fac3394ad while batching the required tied-mtime fix. The strict v2 ID and role-capture expression now lives in V2_BUNDLE_ID_PATTERN and is reused for candidate filtering, per-role grouping, and missing-payload diagnostics. Scoped RuboCop and exact-head QA are clean.

File: benchmarks/bench-node-renderer.rb [hunk 3]
```diff
@@ -95,15 +131,24 @@ def find_all_production_bundles
...
+            else
+              newest_v2_bundles_by_role(bundles_dir, v2_candidates)
+            end
+  bundles.sort_by! do |entry|
+    -bundle_entry_mtime(bundles_dir, entry).to_f
   end
 
   raise "No production bundles found in #{bundles_dir}" if bundles.empty?
 
+  validate_bundle_payloads!(bundles_dir, bundles)
```
[the remark this answers is not in the document]
Comment (justin808) [reply benchmarks/bench-node-renderer.rb]: Addressed the PR-introduced compatibility change in 922d693226c2ea7c305140b58b1eb9e66a609c98: exact payload validation now runs only for canonical v2 discovery, and a regression example proves legacy discovery does not gain eager payload validation. Independent QA also disproved the supporting premise that the legacy live categorizer skips HTTP 410 responses (curl does not use --fail); rsc_bundle? and categorize_bundles are byte-identical to base, so this narrow PR intentionally does not expand into that separate pre-existing behavior. Exact-head QA is CLEAN; focused file 18/0 and all benchmark specs 350/0.

File: benchmarks/bench-node-renderer.rb [hunk 4]
```diff
@@ -77,12 +77,48 @@ def read_password_from_config
...
+  bundles.each do |entry|
+    payload_path = File.join(bundles_dir, entry, "#{entry}.js")
+    next if File.file?(payload_path)
+
+    role = entry[V2_BUNDLE_ID_PATTERN, 1]
+    message = if role
+                "Newest node renderer bundle for role #{role} is missing its payload: #{payload_path}"
+              else
+                "Node renderer bundle #{entry} is missing its payload: #{payload_path}"
+              end
```
[the remark this answers is not in the document]
Comment (justin808) [reply benchmarks/bench-node-renderer.rb:117]: [auto-deferred]

Declined as a no-op at the final-candidate debounce point. The reviewer confirms this is not a bug; retaining the defensive fallback preserves a clear generic error if this private helper is reused, while removing it would restart exact-head CI/review for no behavior benefit. No code or follow-up issue is warranted.

**Comment by justin808** [comment]
## Address-review summary

Scan scope: full PR history; no previous address-review summary existed, so this is equivalent to `check all reviews`.

### Mattered

- Greptile P2 on tied same-role mtimes: fixed in `278f51f1`; exact ties now fail closed with deterministic diagnostics. Replied and resolved.
- Claude finding on eager legacy payload validation: fixed in `922d6932`; fail-closed payload validation is confined to selected v2 artifacts and legacy discovery retains its pre-PR contract. The review's supporting claim that the legacy classifier skipped HTTP 410 responses was disproved against the unchanged base implementation, so no unrelated classifier change was made. Replied and resolved.

### Optional

- Claude DRY nit on duplicated role parsing: fixed in `278f51f1` by using the anchored v2 ID pattern consistently. Replied and resolved.
- Claude current-head nit on the defensive `role.nil?` diagnostic: declined under the final-candidate debounce rule. It is explicitly not a bug, retains a useful generic error if the private helper is reused, and changing it would restart exact-head CI/review for no behavior benefit. Replied and resolved; no follow-up issue.

### Skipped

- The Greptile and three Claude general review summaries were triaged as context; their actionable inline findings are dispositioned above. The remaining mtime-heuristic/shared-source observations do not warrant broader changes in this narrow benchmark-harness fix, and the current-head Claude review found nothing blocking.
- The CodeRabbit walkthrough/checkbox boilerplate and four CI command/status comments are non-actionable operational metadata; no code or reply action was needed.

Deferred-work tracking: none; no follow-up issue was created.

Exact-head evidence: `922d693226c2ea7c305140b58b1eb9e66a609c98`; benchmark specs, full lint/build, CodeQL, CodeRabbit, Claude review, and the required PR gate are terminal and green. All four review threads are resolved.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.

**Comment by justin808** [comment]
### Merge authorization receipt

Maintainer merge approval was given in the current Codex task for exact head 922d693226c2ea7c305140b58b1eb9e66a609c98 against main at f4a1bde1e880b5ea69c08c55f620fbf6c10afe19.

Diff identity (SHA-256 of the exact binary base...head diff): 7b8fa5c7329760a245e92641f1b1789ac65c74960261d38b7b7dd00774ab8bc7.

This authorizes the guarded merge of that exact head. It does not waive the required post-merge unattended 05:00 HST Core, Pro, and Pro Node Renderer upload receipt.
